cleanlab/data_valuation.py

```diff
@@ -45,15 +46,19 @@ def _knn_shapley_score(knn_graph: csr_matrix, labels: np.ndarray, k: int) -> np.
 
 
 def _process_knn_graph_from_features(
-    features: np.ndarray, metric: Optional[str], k: int = 10
+    features: np.ndarray, metric: Optional[Union[str, Callable]], k: int = 10
 ) -> csr_matrix:
     """Calculate the knn graph from the features if it is not provided in the kwargs."""
     if k > len(features):  # Ensure number of neighbors less than number of examples
         raise ValueError(
             f"Number of nearest neighbors k={k} cannot exceed the number of examples N={len(features)} passed into the estimator (knn)."
         )
     if metric == None:
-        metric = "cosine" if features.shape[1] > 3 else "euclidean"
+        metric = (
+            "cosine"
+            if features.shape[1] > 3
+            else "euclidean" if features.shape[0] > 100 else euclidean
```

Comment:
upon further reflection, the limitations of this are unclear to me.

What happens when:

1. the dataset has size 101 and is all identical?

2. the rest of the unit tests that were supposed to be added by now from this GH issue: https://github.com/cleanlab/cleanlab/issues/1055
in scenarios where the dataset is less than or greater than n=100